File: fetch.py
#!/usr/bin/env python3
import sys
import os
import subprocess

def get_os_info():
    kernel = subprocess.check_output("uname -r", shell=True, text=True).strip()
    whoami = subprocess.check_output("whoami", shell=True, text=True).strip()
    arch = subprocess.check_output("uname -m", shell=True, text=True).strip()
    host = subprocess.check_output("cat /etc/hostname", shell=True, text=True).strip()
    shell = subprocess.check_output("echo $SHELL", shell=True, text=True).strip()
    term = subprocess.check_output("echo $TERM", shell=True, text=True).strip()
    cpuinfo = subprocess.check_output("cat /proc/cpuinfo", shell=True, text=True).split("\n")
    osrelease = subprocess.check_output("cat /etc/os-release", shell=True, text=True).split("\n")
    return kernel, whoami, cpuinfo, osrelease, arch, shell, host, term

def parse_os_logo(ID, osf):
    oslogo = [""] * len(osf)
    startRead = False
    i = 0
    line: str = ""
    linedex: int = 0
    for line in osf:
        linedex += 1
        if line == ID:
            startRead = True
            continue
        if line[0] == "#":
            continue
        if line == ":EOF" and startRead:
            break
        if startRead:
            oslogo[i] = line.replace("\\033", '\033')
            i += 1
    while i < len(oslogo) - linedex:
        if oslogo[i] == "":
            oslogo[i] = osf[linedex-2].replace("\\033", '\033')
        i += 1
    return oslogo

def parse_os_release(osrelease):
    ID = ""
    NAME = ""
    for element in osrelease:
        if "ID=" in element and not "BUILD" in element:
            ID = element.replace("ID=", '').strip()
        if "NAME=" in element and not "PRETTY" in element:
            NAME = element.replace("NAME=", '').replace("\"", '').strip()
    return ID, NAME

def parse_shell(shell):
    if "bash" in shell:
        return "bash"
    elif "zsh" in shell:
        return "zsh"
    else:
        return "Other"

def parse_term(term):
    if "kitty" in term:
        return "kitty"
    else:
        return term
    
def main():
    kernel, whoami, cpuinfo, osrelease, arch, shell, host, term = get_os_info()

    ID, NAME = parse_os_release(osrelease)
    shell = parse_shell(shell)
    term = parse_term(term)

    i = 1
    while i < len(sys.argv):
        if sys.argv[i] == "-d" or sys.argv[i] == "--distro":
            ID = sys.argv[i+1]
        i += 1

    osf = open('os.txt', 'r').read().split("\n")
    oslogo = parse_os_logo(ID, osf)

    print()
    print(oslogo[0], "OS:\033[0m", NAME)
    print(oslogo[1], "Kernel:\033[0m", kernel)
    print(oslogo[2], "CPU:\033[0m", arch + " " + cpuinfo[4].replace("model name	: ", '').replace(" CPU", ''))
    print(oslogo[3], "Shell:\033[0m", shell)
    print(oslogo[4], "Host:\033[0m", host)
    print(oslogo[5], "User:\033[0m", whoami)
    print(oslogo[6], "Term:\033[0m", term)
    print("\033[0m")

if __name__ == "__main__":
    main()

File: fetch.lua
function parse_shell(shell)
   if string.find(shell, "bash") ~= nil and string.find(shell, "bash") > 0 then
      return "bash"
   end

   if string.find(shell, "zsh") ~= nil and string.find(shell, "zsh") > 0 then
      return "zsh"
   end   
   
   return shell
end

function parse_term(term)
   if string.find(term, "kitty") ~= nil and string.find(term, "kitty") > 0 then
      return "kitty"
   end
   
   return term
end

function parse_cpuinfo(cpuinfo)

   --parse /proc/cpuinfo file
   local current_line   = ""
   local past_colon     = false
   local cpu_model      = ""
   local read_cpu_model = false
   
   for char in cpuinfo:lines(1) do
      if char == '\n' then
	 current_line = ""
	 past_colon = false
	 read_cpu_model = false
	 goto continue
      end
      if char == ':'  then
	 past_colon = true
	 goto continue
      end

      current_line = current_line .. char

      if current_line == "model name" then read_cpu_model = true end
      
      if read_cpu_model and past_colon then
	 cpu_model = cpu_model .. char
      end

      if current_line == "power management" then break end
      
      ::continue::
   end

   local cpu_model_rebuild = ""
   local cpu_i = string.find(cpu_model, "CPU")
   local i = 1
   for _=0, string.len(cpu_model) do
      if i == cpu_i then i = i + string.len("CPU") + 1 end
      
      cpu_model_rebuild = cpu_model_rebuild .. cpu_model:sub(i, i)
      i = i + 1
   end

   
   return cpu_model_rebuild
end

function sys_info()
   local kernel   = io.popen("uname -r", "r"):read()
   local arch     = io.popen("uname -m", "r"):read()
   local hostname = io.popen("uname -n", "r"):read()
   local whoami   = io.popen("whoami",   "r"):read()
   local shell    = parse_shell(os.getenv("SHELL"))
   local term     = parse_term(os.getenv("TERM"))
   local cpuinfo  = parse_cpuinfo(io.open("/proc/cpuinfo", "r"))
   
   return kernel, arch, whoami, hostname, shell, term, cpuinfo
end

function main()
   local kernel, arch, whoami, hostname, shell, term, cpuinfo = sys_info()
   
   print("Kernel: " .. kernel)
   print("Shell: "  .. shell)
   print("CPU: "    .. arch .. cpuinfo)
   print("Host: "   .. hostname)
   print("User: "   .. whoami)
   print("Term: "   .. term)

end

main()
